skimage/morphology/_skeletonize_3d_cy.pyx.in

```diff
@@ -0,0 +1,430 @@
+                    img[p, r, c] = 0
+                    no_change = False
+
+            if no_change:
+                unchanged_borders += 1
+
+    return np.asarray(img)
+
+
+@cython.boundscheck(False)
+@cython.wraparound(False)
+cdef list _loop_through(pixel_type[:, :, ::1] img,
+                        int curr_border):
+    """Inner loop of compute_thin_image.
+
+    The algorithm of [Lee94]_ proceeds in two steps: (1) six directions are
+    checked for simple border points to remove, and (2) these candidates are
+    sequentially rechecked, see Sec 3 of [Lee94]_ for rationale and discussion.
+
+    This routine implements the first step above: it loops over the image
+    for a given direction and assembles candidates for removal.
+
+    """
+    # This routine looks like it could be nogil, but actually it cannot be,
+    # because of `simple_border_points` being a python list which is being
+    # mutated.
+    cdef:
+        list simple_border_points = []
+        pixel_type neighborhood[27]
+        npy_intp p, r, c
+        bint is_border_pt
+        (npy_intp, npy_intp, npy_intp) point
+
+        # rebind a global name to avoid lookup. The table is filled in
+        # at import time.
+        int[::1] Euler_LUT = LUT
+
+    # loop through the image
+    # NB: each loop is from 1 to size-1: img is padded from all sides 
+    for p in range(1, img.shape[0] - 1):
+        for r in range(1, img.shape[1] - 1):
+            for c in range(1, img.shape[2] - 1):
+
+                # check if pixel is foreground
+                if img[p, r, c] != 1:
+                    continue
+
+                is_border_pt = (curr_border == 1 and img[p, r, c-1] == 0 or  #N
+                                curr_border == 2 and img[p, r, c+1] == 0 or  #S
+                                curr_border == 3 and img[p, r+1, c] == 0 or  #E
+                                curr_border == 4 and img[p, r-1, c] == 0 or  #W
+                                curr_border == 5 and img[p+1, r, c] == 0 or  #U
+                                curr_border == 6 and img[p-1, r, c] == 0)    #B
+                if not is_border_pt:
+                    # current point is not deletable
+                    continue
+
+                get_neighborhood(img, p, r, c, neighborhood)
+
+                # check if (p, r, c) is an endpoint (then it's not deletable.)
+                if is_endpoint(neighborhood):
+                    continue
+
+                # check if point is Euler invariant (condition 1 in [Lee94]_):
+                # if it is not, it's not deletable.
+                if not is_Euler_invariant(neighborhood, Euler_LUT):
+                    continue
+
+                # check if point is simple (i.e., deletion does not
+                # change connectivity in the 3x3x3 neighborhood)
+                # this are conditions 2 and 3 in [Lee94]_
+                if not is_simple_point(neighborhood):
+                    continue
```

Comment:
Minor: imho, these three clauses would be more readable as a single if statement.